### PR TITLE
Added defaultClientID to cluster 

### PR DESCRIPTION
**Pull request by akarukappadath**
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by akarukappadath** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterConcurrencySpec.scala [hunk 1]
```diff
@@ -111,6 +111,17 @@ class ClusterConcurrencySpec extends FreeSpec with LeonardoTestUtils with Parall
         }
       }
     }
+
+    // make sure defaultClientId works
+    "should set a default client id if specified" in {
+      withProject { project => implicit token =>
+        val request = defaultClusterRequest.copy(defaultClientId = Some("this is a client ID"))
+        withNewCluster(project, request = request) { cluster =>
```
Comment (aweng98) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterConcurrencySpec.scala]: can you have param `monitorDelete=false`?

**Comment by aweng98** [comment]
all checks passed. Amazing!

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterConcurrencySpec.scala (hunk 1 above)
Comment (akarukappadath) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterConcurrencySpec.scala]: ok!

**Comment by akarukappadath** [comment]
jenkins retest

File: src/main/resources/jupyter/google_sign_in.js [hunk 2]
```diff
@@ -18,7 +18,7 @@ var googleProject = $(googleProject);
 var clusterName = $(clusterName);
 
 // This is refreshed via postMessage from the client app.
-var googleClientId;
+var googleClientId = $(defaultClientId);
```
Comment (vkumra-broad) [inline src/main/resources/jupyter/google_sign_in.js]: Question: would this maintain backward compatibility with the post message way of doing this?

File: src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20180810_default_client_id.xml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="leonardo" xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+    <changeSet logicalFilePath="leonardo" author="azmina" id="default_client_id">
+        <addColumn tableName="CLUSTER">
+            <column name="defaultClientId" type="varchar(1024)">
+                <constraints nullable="true" />
+            </column>
+        </addColumn>
+    </changeSet>
+</databaseChangeLog>
```
Comment (vkumra-broad) [inline src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20180810_default_client_id.xml]: nit: new line.

File: src/main/resources/swagger/api-docs.yaml [hunk 4]
```diff
@@ -675,6 +676,9 @@ definitions:
       autopauseThreshold:
         type: integer
         description: The number of minutes of idle time to elapse before the cluster is autopaused. A value of 0 is equivalent to autopause being turned off.
+      defaultClientId:
+        type: string
+        description: The Google Client ID to be used if using All of Us or Saturn.
```
Comment (vkumra-broad) [inline src/main/resources/swagger/api-docs.yaml]: Could we make the description more generic?

File: src/main/resources/swagger/api-docs.yaml [hunk 5]
```diff
@@ -754,6 +758,9 @@ definitions:
       autopauseThreshold:
         type: integer
         description: The number of minutes of idle time to elapse before the cluster is autopaused. If autopause is set to false, this value is disregarded. A value of 0 is equivalent to autopause being turned off. If autopause is enabled and this is unset, a system default threshold will be used.
+      defaultClientId:
+        type: string
+        description: The Google Client ID to be used if using All of Us or Saturn.
```
Comment (vkumra-broad) [inline src/main/resources/swagger/api-docs.yaml]: same here.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala [hunk 6]
```diff
@@ -281,7 +286,7 @@ object LeonardoJsonSupport extends SprayJsonSupport with DefaultJsonProtocol {
   implicit val DefaultLabelsFormat = jsonFormat6(DefaultLabels.apply)
 
 
-  implicit val ClusterInitValuesFormat = jsonFormat22(ClusterInitValues.apply)
+ // implicit val ClusterInitValuesFormat = jsonFormat22(ClusterInitValues.apply)
```
Comment (vkumra-broad) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: can be removed if we're not using it.

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 7]
```diff
@@ -555,8 +556,10 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
   }
 
   it should "template google_sign_in.js with config values" in isolatedDbTest {
+
     // Create replacements map
-    val replacements = ClusterInitValues(project, name1, initBucketPath, testClusterRequest, dataprocConfig, clusterFilesConfig, clusterResourcesConfig, proxyConfig, Some(serviceAccountKey), userInfo.userEmail, contentSecurityPolicy).toJson.asJsObject.fields
+    val clusterInit = ClusterInitValues(project, name1, initBucketPath, testClusterRequest, dataprocConfig, clusterFilesConfig, clusterResourcesConfig, proxyConfig, Some(serviceAccountKey), userInfo.userEmail, contentSecurityPolicy)
```
Comment (vkumra-broad) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: I think we should update the test to check here that the defaultClientId is templated correctly in test_google_sign_in.js

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 8]
```diff
@@ -532,7 +532,8 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
 
   it should "template a script using config values" in isolatedDbTest {
     // Create replacements map
-    val replacements = ClusterInitValues(project, name1, initBucketPath, testClusterRequestWithExtensionAndScript, dataprocConfig, clusterFilesConfig, clusterResourcesConfig, proxyConfig, Some(serviceAccountKey), userInfo.userEmail, contentSecurityPolicy).toJson.asJsObject.fields
+    val clusterInit = ClusterInitValues(project, name1, initBucketPath, testClusterRequestWithExtensionAndScript, dataprocConfig, clusterFilesConfig, clusterResourcesConfig, proxyConfig, Some(serviceAccountKey), userInfo.userEmail, contentSecurityPolicy)
+    val replacements: Map[String, String] = clusterInit.getClass.getDeclaredFields.map(_.getName).zip(clusterInit.productIterator.to).toMap.mapValues(_.toString)
```
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: We can use the `toMap` method (mentioned on a comment above) here and below.

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala (hunk 7 above)
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: Can we add `defaultClientId` to `testClusterRequest` (probably via `testClusterRequest.copy(defaultClientId = ...)`) and test it below?

File: src/main/resources/jupyter/google_sign_in.js (hunk 2 above)
Comment (akarukappadath) [reply src/main/resources/jupyter/google_sign_in.js]: Yup! In Firecloud, the postmessage route overwrites the googleClientId var in the receive method below

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 9]
```diff
@@ -675,10 +675,12 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
 
+    //
+    val clusterInit = ClusterInitValues(googleProject, clusterName, initBucketName, clusterRequest, dataprocConfig,
+      clusterFilesConfig, clusterResourcesConfig, proxyConfig, serviceAccountKey, userEmail, contentSecurityPolicy)
+
     // Build a mapping of (name, value) pairs with which to apply templating logic to resources
-    val replacements: Map[String, JsValue] = ClusterInitValues(googleProject, clusterName, initBucketName, clusterRequest, dataprocConfig,
-      clusterFilesConfig, clusterResourcesConfig, proxyConfig, serviceAccountKey, userEmail, contentSecurityPolicy
-    ).toJson.asJsObject.fields
+    val replacements: Map[String, String] = clusterInit.getClass.getDeclaredFields.map(_.getName).zip(clusterInit.productIterator.to).toMap.mapValues(_.toString)
```
[the remark this answers is not in the document]
Comment (akarukappadath) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: great idea!

[the remark this answers is not in the document]
Comment (akarukappadath) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: For testing this method, should I be checking a couple of the entries of the map, or every single one?

**Comment by akarukappadath** [comment]
Manually tested using the Saturn UI

**Pull request merged**